<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_101) on Mon Oct 17 16:12:02 NZDT 2016 -->
<title>AlignmentLayout (Geneious Public API)</title>
<meta name="date" content="2016-10-17">
<link rel="stylesheet" type="text/css" href="../../../../../../stylesheet.css" title="Style">
<script type="text/javascript" src="../../../../../../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="AlignmentLayout (Geneious Public API)";
        }
    }
    catch(err) {
    }
//-->
var methods = {"i0":10,"i1":9,"i2":10,"i3":10,"i4":10,"i5":10,"i6":10,"i7":10,"i8":10,"i9":10,"i10":10,"i11":10};
var tabs = {65535:["t0","All Methods"],1:["t1","Static Methods"],2:["t2","Instance Methods"],8:["t4","Concrete Methods"]};
var altColor = "altColor";
var rowColor = "rowColor";
var tableTab = "tableTab";
var activeTableTab = "activeTableTab";
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/AlignmentLayout.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../index-all.html">Index</a></li>
<li><a href="../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev&nbsp;Class</li>
<li><a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AminoAcidSequenceDocument.html" title="interface in com.biomatters.geneious.publicapi.documents.sequence"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../../../index.html?com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html" target="_top">Frames</a></li>
<li><a href="AlignmentLayout.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">com.biomatters.geneious.publicapi.documents.sequence</div>
<h2 title="Class AlignmentLayout" class="title">Class AlignmentLayout</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li><a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/Object.html?is-external=true" title="class or interface in java.lang">java.lang.Object</a></li>
<li>
<ul class="inheritance">
<li>com.biomatters.geneious.publicapi.documents.sequence.AlignmentLayout</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<dl>
<dt>All Implemented Interfaces:</dt>
<dd><a href="../../../../../../com/biomatters/geneious/publicapi/documents/XMLSerializable.OldVersionCompatible.html" title="interface in com.biomatters.geneious.publicapi.documents">XMLSerializable.OldVersionCompatible</a></dd>
</dl>
<hr>
<br>
<pre>public final class <span class="typeNameLabel">AlignmentLayout</span>
extends <a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/Object.html?is-external=true" title="class or interface in java.lang">Object</a>
implements <a href="../../../../../../com/biomatters/geneious/publicapi/documents/XMLSerializable.OldVersionCompatible.html" title="interface in com.biomatters.geneious.publicapi.documents">XMLSerializable.OldVersionCompatible</a></pre>
<div class="block">Provides information about how to layout of sequences (excluding the reference sequence) for rendering in an alignment or contig.
 Each sequence is assigned to a row number such that all the sequences in a row do not overlap
 each other and have a minimum separation. Paired reads can optionally be forced to appear
 in the same row as their mate, with no sequences in between.
 <p/>
 In order to support very large alignments or contigs, the data is never all loaded into memory at once,
 but is instead loaded on demand (with caching)
 <p/>
 An AlignmentLayout can be obtained from a <a href="../../../../../../com/biomatters/geneious/publicapi/implementations/DefaultAlignmentDocument.html" title="class in com.biomatters.geneious.publicapi.implementations"><code>DefaultAlignmentDocument</code></a> via
 <a href="../../../../../../com/biomatters/geneious/publicapi/implementations/DefaultAlignmentDocument.html#getAlignmentDataForSequencesNotInMemory--"><code>DefaultAlignmentDocument.getAlignmentDataForSequencesNotInMemory()</code></a>.<a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/SequenceListOnDisk.AlignmentData.html#getLinkAllPairsLayout--"><code>getLinkAllPairsLayout()</code></a> or
 <a href="../../../../../../com/biomatters/geneious/publicapi/implementations/DefaultAlignmentDocument.html#getAlignmentDataForSequencesNotInMemory--"><code>DefaultAlignmentDocument.getAlignmentDataForSequencesNotInMemory()</code></a>.<a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/SequenceListOnDisk.AlignmentData.html#getLinkNearbyPairsLayout--"><code>getLinkNearbyPairsLayout()</code></a> or
 <a href="../../../../../../com/biomatters/geneious/publicapi/implementations/DefaultAlignmentDocument.html#getAlignmentDataForSequencesNotInMemory--"><code>DefaultAlignmentDocument.getAlignmentDataForSequencesNotInMemory()</code></a>.<a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/SequenceListOnDisk.AlignmentData.html#getLinkNoPairsLayout--"><code>getLinkNoPairsLayout()</code></a>
 Code should not construct an AlignmentLayout directly. Instead, AlignmentLayouts are constructed as part of creating an alignment via a <a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/SequenceListOnDisk.Builder.html" title="class in com.biomatters.geneious.publicapi.documents.sequence"><code>SequenceListOnDisk.Builder</code></a>
 <p/>
 Unless you are writing your own contig viewer, you can probably ignore this class.</div>
<dl>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>API 4.40 (Geneious 5.4.0)</dd>
</dl>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#AlignmentLayout-org.jdom.Element-">AlignmentLayout</a></span>(org.jdom.Element&nbsp;element)</code>
<div class="block">Recreates an AlignmentLayout from XML.</div>
</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method.summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span id="t0" class="activeTableTab"><span>All Methods</span><span class="tabEnd">&nbsp;</span></span><span id="t1" class="tableTab"><span><a href="javascript:show(1);">Static Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t2" class="tableTab"><span><a href="javascript:show(2);">Instance Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t4" class="tableTab"><span><a href="javascript:show(8);">Concrete Methods</a></span><span class="tabEnd">&nbsp;</span></span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr id="i0" class="altColor">
<td class="colFirst"><code>protected void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#finalize--">finalize</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i1" class="rowColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#getMaximumColumnsCached--">getMaximumColumnsCached</a></span>()</code>
<div class="block">Returns the maximum number of columns in the alignment which will be cached in memory at one time.</div>
</td>
</tr>
<tr id="i2" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#getNumberOfColumnsInAlignment--">getNumberOfColumnsInAlignment</a></span>()</code>
<div class="block">Returns the number of columns in the alignment</div>
</td>
</tr>
<tr id="i3" class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#getNumberOfRows--">getNumberOfRows</a></span>()</code>
<div class="block">Returns the total number of rows used for laying out sequences</div>
</td>
</tr>
<tr id="i4" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#getNumberOfSequences--">getNumberOfSequences</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i5" class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#getRowNumber-int-">getRowNumber</a></span>(int&nbsp;sequenceNumber)</code>
<div class="block">Get the row number that the given sequence has been lain out in.</div>
</td>
</tr>
<tr id="i6" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#getSequenceInRowNearestToColumn-int-int-">getSequenceInRowNearestToColumn</a></span>(int&nbsp;rowNumber,
                               int&nbsp;columnIndex)</code>
<div class="block">For a given row, gets the sequence covering this column index, or if there isn't a sequence
 covering this column, returns the nearest sequence to it.</div>
</td>
</tr>
<tr id="i7" class="rowColor">
<td class="colFirst"><code><a href="../../../../../../com/biomatters/geneious/publicapi/utilities/IntList.html" title="class in com.biomatters.geneious.publicapi.utilities">IntList</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#getSequencesInRowIntersectingRange-int-int-int-">getSequencesInRowIntersectingRange</a></span>(int&nbsp;rowNumber,
                                  int&nbsp;firstColumn,
                                  int&nbsp;lastColumnInclusive)</code>
<div class="block">Gets all sequences in the given row that intersect with the given range of columns</div>
</td>
</tr>
<tr id="i8" class="altColor">
<td class="colFirst"><code><a href="../../../../../../com/biomatters/geneious/publicapi/plugin/Geneious.MajorVersion.html" title="enum in com.biomatters.geneious.publicapi.plugin">Geneious.MajorVersion</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#getVersionSupport-com.biomatters.geneious.publicapi.documents.XMLSerializable.VersionSupportType-">getVersionSupport</a></span>(<a href="../../../../../../com/biomatters/geneious/publicapi/documents/XMLSerializable.VersionSupportType.html" title="enum in com.biomatters.geneious.publicapi.documents">XMLSerializable.VersionSupportType</a>&nbsp;versionType)</code>
<div class="block">Gets either the latest version this class can serialize to via <a href="../../../../../../com/biomatters/geneious/publicapi/documents/XMLSerializable.OldVersionCompatible.html#toXML-com.biomatters.geneious.publicapi.plugin.Geneious.MajorVersion-jebl.util.ProgressListener-"><code>XMLSerializable.OldVersionCompatible.toXML(com.biomatters.geneious.publicapi.plugin.Geneious.MajorVersion, jebl.util.ProgressListener)</code></a>
 or the most recent version in which the XML returned from <a href="../../../../../../com/biomatters/geneious/publicapi/documents/XMLSerializable.OldVersionCompatible.html#toXML-com.biomatters.geneious.publicapi.plugin.Geneious.MajorVersion-jebl.util.ProgressListener-"><code>XMLSerializable.OldVersionCompatible.toXML(com.biomatters.geneious.publicapi.plugin.Geneious.MajorVersion, jebl.util.ProgressListener)</code></a> last changed.</div>
</td>
</tr>
<tr id="i9" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#isLinkPairedReadsInRow-int-">isLinkPairedReadsInRow</a></span>(int&nbsp;rowNumber)</code>
<div class="block">Returns true if pairs reads in the given row number are linked (i.e.</div>
</td>
</tr>
<tr id="i10" class="altColor">
<td class="colFirst"><code>org.jdom.Element</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#toXML-com.biomatters.geneious.publicapi.plugin.Geneious.MajorVersion-jebl.util.ProgressListener-">toXML</a></span>(<a href="../../../../../../com/biomatters/geneious/publicapi/plugin/Geneious.MajorVersion.html" title="enum in com.biomatters.geneious.publicapi.plugin">Geneious.MajorVersion</a>&nbsp;majorVersion,
     <a href="http://jebl.sourceforge.net/doc/api/jebl/util/ProgressListener.html?is-external=true" title="class or interface in jebl.util">ProgressListener</a>&nbsp;progressListener)</code>
<div class="block">Serializes this class to XML format, potentially to a format readable by an earlier version of Geneious.</div>
</td>
</tr>
<tr id="i11" class="rowColor">
<td class="colFirst"><code>org.jdom.Element</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#toXML-java.lang.String-">toXML</a></span>(<a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;elementName)</code>
<div class="block">Converts this layout to XML for use with <a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#AlignmentLayout-org.jdom.Element-"><code>AlignmentLayout(org.jdom.Element)</code></a></div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.<a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/Object.html?is-external=true" title="class or interface in java.lang">Object</a></h3>
<code><a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/Object.html?is-external=true#clone--" title="class or interface in java.lang">clone</a>, <a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/Object.html?is-external=true#equals-java.lang.Object-" title="class or interface in java.lang">equals</a>, <a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/Object.html?is-external=true#getClass--" title="class or interface in java.lang">getClass</a>, <a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/Object.html?is-external=true#hashCode--" title="class or interface in java.lang">hashCode</a>, <a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/Object.html?is-external=true#notify--" title="class or interface in java.lang">notify</a>, <a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/Object.html?is-external=true#notifyAll--" title="class or interface in java.lang">notifyAll</a>, <a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/Object.html?is-external=true#toString--" title="class or interface in java.lang">toString</a>, <a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/Object.html?is-external=true#wait--" title="class or interface in java.lang">wait</a>, <a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/Object.html?is-external=true#wait-long-" title="class or interface in java.lang">wait</a>, <a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/Object.html?is-external=true#wait-long-int-" title="class or interface in java.lang">wait</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="AlignmentLayout-org.jdom.Element-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>AlignmentLayout</h4>
<pre>public&nbsp;AlignmentLayout(org.jdom.Element&nbsp;element)
                throws <a href="../../../../../../com/biomatters/geneious/publicapi/documents/XMLSerializationException.html" title="class in com.biomatters.geneious.publicapi.documents">XMLSerializationException</a></pre>
<div class="block">Recreates an AlignmentLayout from XML.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>element</code> - an element returned from <a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#toXML-java.lang.String-"><code>toXML(String)</code></a></dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code><a href="../../../../../../com/biomatters/geneious/publicapi/documents/XMLSerializationException.html" title="class in com.biomatters.geneious.publicapi.documents">XMLSerializationException</a></code> - if we can't deserialize because the XML is invalid</dd>
</dl>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method.detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="getVersionSupport-com.biomatters.geneious.publicapi.documents.XMLSerializable.VersionSupportType-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getVersionSupport</h4>
<pre>public&nbsp;<a href="../../../../../../com/biomatters/geneious/publicapi/plugin/Geneious.MajorVersion.html" title="enum in com.biomatters.geneious.publicapi.plugin">Geneious.MajorVersion</a>&nbsp;getVersionSupport(<a href="../../../../../../com/biomatters/geneious/publicapi/documents/XMLSerializable.VersionSupportType.html" title="enum in com.biomatters.geneious.publicapi.documents">XMLSerializable.VersionSupportType</a>&nbsp;versionType)</pre>
<div class="block"><span class="descfrmTypeLabel">Description copied from interface:&nbsp;<code><a href="../../../../../../com/biomatters/geneious/publicapi/documents/XMLSerializable.OldVersionCompatible.html#getVersionSupport-com.biomatters.geneious.publicapi.documents.XMLSerializable.VersionSupportType-">XMLSerializable.OldVersionCompatible</a></code></span></div>
<div class="block">Gets either the latest version this class can serialize to via <a href="../../../../../../com/biomatters/geneious/publicapi/documents/XMLSerializable.OldVersionCompatible.html#toXML-com.biomatters.geneious.publicapi.plugin.Geneious.MajorVersion-jebl.util.ProgressListener-"><code>XMLSerializable.OldVersionCompatible.toXML(com.biomatters.geneious.publicapi.plugin.Geneious.MajorVersion, jebl.util.ProgressListener)</code></a>
 or the most recent version in which the XML returned from <a href="../../../../../../com/biomatters/geneious/publicapi/documents/XMLSerializable.OldVersionCompatible.html#toXML-com.biomatters.geneious.publicapi.plugin.Geneious.MajorVersion-jebl.util.ProgressListener-"><code>XMLSerializable.OldVersionCompatible.toXML(com.biomatters.geneious.publicapi.plugin.Geneious.MajorVersion, jebl.util.ProgressListener)</code></a> last changed.
 <p/>
 Example implementation:
 <pre><code>
 switch (versionType) {
   case FormatLastExtended:
     //added a new tag to xml in 10.0 that can be safely ignored by older versions
     return Geneious.MajorVersion.Version10_0;
   case FormatLastChanged:
     //renamed a tag in 9.1 meaning that older versions can no longer read the xml from 9.1
     return Geneious.MajorVersion.Version9_1;
   case OldestVersionSerializableTo:
     //this document class can export xml that is compatible with 9.0
     return Geneious.MajorVersion.Version9_0;
   default:
     throw new IllegalArgumentException("Unrecognized VersionSupportType");
 }
 </code></pre></div>
<dl>
<dt><span class="overrideSpecifyLabel">Specified by:</span></dt>
<dd><code><a href="../../../../../../com/biomatters/geneious/publicapi/documents/XMLSerializable.OldVersionCompatible.html#getVersionSupport-com.biomatters.geneious.publicapi.documents.XMLSerializable.VersionSupportType-">getVersionSupport</a></code>&nbsp;in interface&nbsp;<code><a href="../../../../../../com/biomatters/geneious/publicapi/documents/XMLSerializable.OldVersionCompatible.html" title="interface in com.biomatters.geneious.publicapi.documents">XMLSerializable.OldVersionCompatible</a></code></dd>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>versionType</code> - specifies which version property to return</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the major version of Geneious corresponding to the versionType for this classes toXML method.</dd>
</dl>
</li>
</ul>
<a name="toXML-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>toXML</h4>
<pre>public&nbsp;org.jdom.Element&nbsp;toXML(<a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;elementName)</pre>
<div class="block">Converts this layout to XML for use with <a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#AlignmentLayout-org.jdom.Element-"><code>AlignmentLayout(org.jdom.Element)</code></a></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>elementName</code> - the name of the element to return</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>this layout as XML. Note that this uses <a href="../../../../../../com/biomatters/geneious/publicapi/documents/PluginDocument.html#FILE_DATA_ATTRIBUTE_NAME"><code>PluginDocument.FILE_DATA_ATTRIBUTE_NAME</code></a></dd>
</dl>
</li>
</ul>
<a name="toXML-com.biomatters.geneious.publicapi.plugin.Geneious.MajorVersion-jebl.util.ProgressListener-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>toXML</h4>
<pre>public&nbsp;org.jdom.Element&nbsp;toXML(<a href="../../../../../../com/biomatters/geneious/publicapi/plugin/Geneious.MajorVersion.html" title="enum in com.biomatters.geneious.publicapi.plugin">Geneious.MajorVersion</a>&nbsp;majorVersion,
                              <a href="http://jebl.sourceforge.net/doc/api/jebl/util/ProgressListener.html?is-external=true" title="class or interface in jebl.util">ProgressListener</a>&nbsp;progressListener)</pre>
<div class="block"><span class="descfrmTypeLabel">Description copied from interface:&nbsp;<code><a href="../../../../../../com/biomatters/geneious/publicapi/documents/XMLSerializable.OldVersionCompatible.html#toXML-com.biomatters.geneious.publicapi.plugin.Geneious.MajorVersion-jebl.util.ProgressListener-">XMLSerializable.OldVersionCompatible</a></code></span></div>
<div class="block">Serializes this class to XML format, potentially to a format readable by an earlier version of Geneious. It is acceptable for the XML to include unnecessary tags that will be ignored by the earlier version. For
 example if the implementation has only extended the XML since the earlier version, then the XML returned may be identical to the XML returned for the current version.<p/>
 See <a href="../../../../../../com/biomatters/geneious/publicapi/documents/XMLSerializable.html#toXML--"><code>XMLSerializable.toXML()</code></a> for a more
 detailed description of what it means to serialize to XML.
 <p/>
 All classes that implement this method must also implement <a href="../../../../../../com/biomatters/geneious/publicapi/documents/XMLSerializable.html#toXML--"><code>XMLSerializable.toXML()</code></a> and should delegate back to
 this method using <a href="../../../../../../com/biomatters/geneious/publicapi/plugin/Geneious.html#getMajorVersion--"><code>Geneious.getMajorVersion()</code></a> and <a href="http://jebl.sourceforge.net/doc/api/jebl/util/ProgressListener.html?is-external=true#EMPTY" title="class or interface in jebl.util"><code>ProgressListener.EMPTY</code></a> as parameters.</div>
<dl>
<dt><span class="overrideSpecifyLabel">Specified by:</span></dt>
<dd><code><a href="../../../../../../com/biomatters/geneious/publicapi/documents/XMLSerializable.OldVersionCompatible.html#toXML-com.biomatters.geneious.publicapi.plugin.Geneious.MajorVersion-jebl.util.ProgressListener-">toXML</a></code>&nbsp;in interface&nbsp;<code><a href="../../../../../../com/biomatters/geneious/publicapi/documents/XMLSerializable.OldVersionCompatible.html" title="interface in com.biomatters.geneious.publicapi.documents">XMLSerializable.OldVersionCompatible</a></code></dd>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>majorVersion</code> - the major version of Geneious to serialize to XML for. For example "6.0" but not "6.0.0". This must be a version
 returned greater or equal to a version returned from
 <a href="../../../../../../com/biomatters/geneious/publicapi/documents/XMLSerializable.OldVersionCompatible.html#getVersionSupport-com.biomatters.geneious.publicapi.documents.XMLSerializable.VersionSupportType-"><code>getVersionSupport</code></a>(<a href="../../../../../../com/biomatters/geneious/publicapi/documents/XMLSerializable.VersionSupportType.html#OldestVersionSerializableTo"><code>VersionSupportType.OldestVersionSerializableTo</code></a>)
 and must never be greater than the current version (<a href="../../../../../../com/biomatters/geneious/publicapi/plugin/Geneious.html#getMajorVersion--"><code>Geneious.getMajorVersion()</code></a>)</dd>
<dd><code>progressListener</code> - for reporting progress and cancelling</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>object encoded as a JDOM element</dd>
</dl>
</li>
</ul>
<a name="getNumberOfColumnsInAlignment--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getNumberOfColumnsInAlignment</h4>
<pre>public&nbsp;int&nbsp;getNumberOfColumnsInAlignment()</pre>
<div class="block">Returns the number of columns in the alignment</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the number of columns in the alignment</dd>
</dl>
</li>
</ul>
<a name="getNumberOfSequences--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getNumberOfSequences</h4>
<pre>public&nbsp;int&nbsp;getNumberOfSequences()</pre>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the number of sequences in this layout</dd>
</dl>
</li>
</ul>
<a name="getNumberOfRows--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getNumberOfRows</h4>
<pre>public&nbsp;int&nbsp;getNumberOfRows()</pre>
<div class="block">Returns the total number of rows used for laying out sequences</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the total number of rows used for laying out sequences</dd>
</dl>
</li>
</ul>
<a name="isLinkPairedReadsInRow-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isLinkPairedReadsInRow</h4>
<pre>public&nbsp;boolean&nbsp;isLinkPairedReadsInRow(int&nbsp;rowNumber)</pre>
<div class="block">Returns true if pairs reads in the given row number are linked (i.e. nothing is laid out between them)</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>rowNumber</code> - the 0-based row number (from 0 to <a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#getNumberOfRows--"><code>getNumberOfRows()</code></a></dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>true if pairs reads in the given row number are linked (i.e. nothing is laid out between them)</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code><a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/IndexOutOfBoundsException.html?is-external=true" title="class or interface in java.lang">IndexOutOfBoundsException</a></code> - if rowNumber is not between 0 and <a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#getNumberOfRows--"><code>getNumberOfRows()</code></a>-1 inclusive.</dd>
</dl>
</li>
</ul>
<a name="getRowNumber-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getRowNumber</h4>
<pre>public&nbsp;int&nbsp;getRowNumber(int&nbsp;sequenceNumber)</pre>
<div class="block">Get the row number that the given sequence has been lain out in.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>sequenceNumber</code> - the 0-based sequence number</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the 0-based row number (will be between 0 and <a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#getNumberOfRows--"><code>getNumberOfRows()</code></a>-1 inclusive)</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code><a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/IndexOutOfBoundsException.html?is-external=true" title="class or interface in java.lang">IndexOutOfBoundsException</a></code> - if sequenceNumber is not between 0 and <a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#getNumberOfSequences--"><code>getNumberOfSequences()</code></a>-1 inclusive.</dd>
</dl>
</li>
</ul>
<a name="getSequenceInRowNearestToColumn-int-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getSequenceInRowNearestToColumn</h4>
<pre>public&nbsp;int&nbsp;getSequenceInRowNearestToColumn(int&nbsp;rowNumber,
                                           int&nbsp;columnIndex)</pre>
<div class="block">For a given row, gets the sequence covering this column index, or if there isn't a sequence
 covering this column, returns the nearest sequence to it.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>rowNumber</code> - the 0-based row number</dd>
<dd><code>columnIndex</code> - the 0-based index of the column. If this is out of the range 0-><a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#getNumberOfColumnsInAlignment--"><code>getNumberOfColumnsInAlignment()</code></a> it will be clipped to that range.</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the nearest sequence. Every row has at least 1 sequence so this method will always return a valid sequence number.</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code><a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/IndexOutOfBoundsException.html?is-external=true" title="class or interface in java.lang">IndexOutOfBoundsException</a></code> - if rowNumber is not between 0 and <a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#getNumberOfRows--"><code>getNumberOfRows()</code></a>-1 inclusive.</dd>
</dl>
</li>
</ul>
<a name="getSequencesInRowIntersectingRange-int-int-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getSequencesInRowIntersectingRange</h4>
<pre>public&nbsp;<a href="../../../../../../com/biomatters/geneious/publicapi/utilities/IntList.html" title="class in com.biomatters.geneious.publicapi.utilities">IntList</a>&nbsp;getSequencesInRowIntersectingRange(int&nbsp;rowNumber,
                                                  int&nbsp;firstColumn,
                                                  int&nbsp;lastColumnInclusive)</pre>
<div class="block">Gets all sequences in the given row that intersect with the given range of columns</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>rowNumber</code> - the 0-based row number</dd>
<dd><code>firstColumn</code> - the 0-based index of the first column. If this is out of the range 0-><a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#getNumberOfColumnsInAlignment--"><code>getNumberOfColumnsInAlignment()</code></a> it will be clipped to that range.</dd>
<dd><code>lastColumnInclusive</code> - the 0-based index (inclusive) of the last column</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>all sequences that intersect the given range of columns.</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code><a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/IndexOutOfBoundsException.html?is-external=true" title="class or interface in java.lang">IndexOutOfBoundsException</a></code> - if rowNumber is not between 0 and <a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html#getNumberOfRows--"><code>getNumberOfRows()</code></a>-1 inclusive.</dd>
</dl>
</li>
</ul>
<a name="finalize--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>finalize</h4>
<pre>protected&nbsp;void&nbsp;finalize()
                 throws <a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/Throwable.html?is-external=true" title="class or interface in java.lang">Throwable</a></pre>
<dl>
<dt><span class="overrideSpecifyLabel">Overrides:</span></dt>
<dd><code><a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/Object.html?is-external=true#finalize--" title="class or interface in java.lang">finalize</a></code>&nbsp;in class&nbsp;<code><a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/Object.html?is-external=true" title="class or interface in java.lang">Object</a></code></dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code><a href="http://java.sun.com/j2se/1.5.0/docs/api/java/lang/Throwable.html?is-external=true" title="class or interface in java.lang">Throwable</a></code></dd>
</dl>
</li>
</ul>
<a name="getMaximumColumnsCached--">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>getMaximumColumnsCached</h4>
<pre>public static&nbsp;int&nbsp;getMaximumColumnsCached()</pre>
<div class="block">Returns the maximum number of columns in the alignment which will be cached in memory at one time. A viewer on this data should not try viewing data when more than this many columns are visible or it will hit cache throttling performance issues.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the maximum number of columns in the alignment which will be cached in memory at one time. A viewer on this data should not try viewing data when more than this many columns are visible or it will hit cache throttling performance issues</dd>
</dl>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/AlignmentLayout.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../index-all.html">Index</a></li>
<li><a href="../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev&nbsp;Class</li>
<li><a href="../../../../../../com/biomatters/geneious/publicapi/documents/sequence/AminoAcidSequenceDocument.html" title="interface in com.biomatters.geneious.publicapi.documents.sequence"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../../../index.html?com/biomatters/geneious/publicapi/documents/sequence/AlignmentLayout.html" target="_top">Frames</a></li>
<li><a href="AlignmentLayout.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
